crates/red_knot/src/module.rs

```diff
@@ -0,0 +1,708 @@
+        let TestCase {
+            mut resolver,
+            files,
+            src,
+            temp_dir: _temp_dir,
+            ..
+        } = create_resolver()?;
+
+        let foo = src.path().join("foo");
+        let bar = foo.join("bar");
+        let baz = bar.join("baz.py");
+
+        std::fs::create_dir_all(&bar)?;
+        std::fs::write(foo.join("__init__.py"), "")?;
+        std::fs::write(bar.join("__init__.py"), "")?;
+        std::fs::write(&baz, "print('Hello, world!')")?;
+
+        let baz_id = resolver.resolve(ModuleName::new("foo.bar.baz"));
+
+        assert!(baz_id.is_some());
+
+        let baz_module = resolver.module(baz_id.unwrap());
+
+        assert_eq!(&src, baz_module.path().root());
+        assert_eq!(&baz, &*files.path(baz_module.path().file()));
+
+        assert_eq!(baz_id, resolver.resolve_path(&baz));
+
+        Ok(())
+    }
+
+    #[test]
+    fn folder_without_init_py() -> std::io::Result<()> {
+        let mut test_case = create_resolver()?;
+
+        let foo = test_case.src.path().join("foo");
+        // The folder `bar` has no `__init__.py`
+        let bar = foo.join("bar");
+        let baz = bar.join("baz.py");
+
+        std::fs::create_dir_all(&bar)?;
+        std::fs::write(foo.join("__init__.py"), "")?;
+        std::fs::write(&baz, "print('Hello, world!')")?;
+
+        let baz_id = test_case.resolver.resolve(ModuleName::new("foo.bar.baz"));
+
+        assert_eq!(None, baz_id);
+        assert_eq!(None, test_case.resolver.resolve_path(&baz));
+
+        Ok(())
+    }
+
+    #[test]
+    fn module_search_path_priority() -> std::io::Result<()> {
+        let TestCase {
+            mut resolver,
+            src,
+            site_packages,
+            files,
+            temp_dir: _temp_dir,
+            ..
+        } = create_resolver()?;
+
+        let foo_src = src.path().join("foo.py");
+        let foo_site_packages = site_packages.path().join("foo.py");
+
+        std::fs::write(&foo_src, "")?;
+        std::fs::write(&foo_site_packages, "")?;
+
+        let foo_id = resolver.resolve(ModuleName::new("foo"));
+
+        assert!(foo_id.is_some());
+
+        let foo_module = resolver.module(foo_id.unwrap());
+
+        assert_eq!(&src, foo_module.path().root());
+        assert_eq!(&foo_src, &*files.path(foo_module.path().file()));
+
+        assert_eq!(foo_id, resolver.resolve_path(&foo_src));
+        assert_eq!(None, resolver.resolve_path(&foo_site_packages));
+
+        Ok(())
+    }
+
+    #[test]
+    #[cfg(target_family = "unix")]
+    fn symlink() -> std::io::Result<()> {
+        let TestCase {
+            mut resolver,
+            files,
+            src,
+            temp_dir: _temp_dir,
+            ..
+        } = create_resolver()?;
+
+        let foo = src.path().join("foo.py");
+        let bar = src.path().join("bar.py");
+
+        std::fs::write(&foo, "")?;
+        std::os::unix::fs::symlink(&foo, src.path().join("bar.py"))?;
```

Comment:
```suggestion
        std::os::unix::fs::symlink(&foo, bar.clone())?;
```

Reply:
Thanks. No need to clone, we can pass a reference `&bar`